src/libraries/System.Reflection.Emit/tests/TypeBuilder/TypeBuilderDefineField.cs

```diff
@@ -60,6 +60,144 @@ public void DefineField(string name, Type fieldType, FieldAttributes attributes,
+                }
+            }
+        }
+
+        [Fact]
+        public void DefineField_NameCollision()
+        {
+            TypeBuilder type = Helpers.DynamicType(TypeAttributes.Public);
+            FieldBuilder field1 = type.DefineField("FieldName", typeof(int), FieldAttributes.Public);
+            FieldBuilder field2 = type.DefineField("FieldName", typeof(int), FieldAttributes.Public);
+            FieldBuilder field3 = type.DefineField("FieldName", typeof(string), FieldAttributes.Public);
+            FieldBuilder field4 = type.DefineField("FieldName", typeof(string), FieldAttributes.Public);
+            Type createdType = type.CreateType();
+
+            FieldInfo[] fields = createdType.GetFields(Helpers.AllFlags);
+            Assert.Equal(4, fields.Length);
+            Assert.Equal("FieldName", fields[0].Name);
+            Assert.Equal("FieldName", fields[1].Name);
+            Assert.Equal("FieldName", fields[2].Name);
+            Assert.Equal("FieldName", fields[3].Name);
+            Assert.Throws<AmbiguousMatchException>(() => createdType.GetField("FieldName", Helpers.AllFlags));
+        }
+
+        [ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsNotBrowser))]
+        public void DefineField_NameCollisionPersistedAssembly()
+        {
+            PersistedAssemblyBuilder ab = AssemblySaveTools.PopulateAssemblyBuilderAndTypeBuilder(out TypeBuilder type);
+            FieldBuilder field1 = type.DefineField("FieldName", typeof(int), FieldAttributes.Public);
+            FieldBuilder field2 = type.DefineField("FieldName", typeof(int), FieldAttributes.Public);
+            FieldBuilder field3 = type.DefineField("FieldName", typeof(string), FieldAttributes.Public);
+            type.CreateType();
+
+            using (var stream = new MemoryStream())
+            using (MetadataLoadContext mlc = new MetadataLoadContext(new CoreMetadataAssemblyResolver()))
+            {
+                ab.Save(stream);
+                Type createdType = mlc.LoadFromStream(stream).GetType("MyType");
+                FieldInfo[] fields = createdType.GetFields(Helpers.AllFlags);
+                Assert.Equal(3, fields.Length);
+                Assert.Equal("FieldName", fields[0].Name);
+                Assert.Equal("FieldName", fields[1].Name);
+                Assert.Throws<AmbiguousMatchException>(() => createdType.GetField("FieldName", Helpers.AllFlags));
+            }
+        }
+
+        [Fact]
+        [ActiveIssue("https://github.com/dotnet/runtime/issues/2389", TestRuntimes.Mono)]
+        public void DefineField_65536Fields()
+        {
+            TypeBuilder type = Helpers.DynamicType(TypeAttributes.Public);
+            for (int i = 0; i < 65536; i++)
+            {
+                type.DefineField($"F_{i}", typeof(int), FieldAttributes.Public);
+            }
+
+            // System.TypeLoadException : Internal limitation: too many fields.
```

Comment:
nit: this isn't really too many fields but an edge case in the ECMA spec when we switch from 2 to 4 bytes to represent the field number.

Reply:
Yes, but I think this is a bit different than that roslyn bug, for reflection TypeLoader it throws whenever the method count is over 65536 with this message